### ISSUE TITLE
Bad encodation of IPv6 URL cause a WebSocket url failed to construct

### ISSUE THREAD

**Issue by petregeorgee** [vaadin/framework#11458]
Hello guys,

* Vaadin Framework version: 7.7.16 ( I know that the support for this version has ended, but can be fixed in a newer version so I could migrate the app)

* Browsers: Chrome, Firefox, Edge, etc

* Description of the bug
I am trying to configure my Vaadin Application to support IPv6 and somehow the URL is encoded incorrectly.

For example: If the ip is: fd71:2d5a:e15a:1ed3::1, on browser you access it like this: http://[fd71:2d5a:e15a:1ed3::1].

The problem is that the character '[' is replaced with '%5B' and the character ']' is replaced with '%5D'.

This encoding cause a "Failed to construct WebSocket: The URL is invalid; JavaScriptException [SyntaxError]" as the picture is describing.
![websocket error](https://user-images.githubusercontent.com/17829580/52844120-92933400-310c-11e9-8303-3a3200897068.png)

The problem is in vaadin-push library, vaadinPush.debug.js file. 

The method is called _buildWebSocketUrl

`function _buildWebSocketUrl() {
                return _attachHeaders(_request, atmosphere.util.getAbsoluteURL(_request.webSocketUrl || _request.url)).replace(/^http/, "ws");
  }`

which calls getAbsoluteURL method, which calls 'encodeURI' and 'decodeURI' functions.

* Expected behavior
We have to skip those [] characters from the encoding.

**Comment by TatuLund** [comment]
Is this the same issue as documented here: https://github.com/Atmosphere/atmosphere-javascript/issues/228 ?

**Comment by petregeorgee** [comment]
Well, almost the same issue, because my problem is on vaadin-push 7.7.16
jar and is replicated on all browsers(Chrome, Firefox and Edge), not only
in Firefox.

Why I opened the issue on vaadin framework? Because I needed to know if I
found the real problem. For that, I've edited locally that JavaScript
file(from VAADIN-PUSH jar) by replaced the %5B and %5d with [ and ]. With
these modifications, my app is running just perfect.

How can we solve this issue? I can open a pull request for this.

Thank you,
George

**Comment by TatuLund** [comment]
> Why I opened the issue on vaadin framework? Because I needed to know if I found the real problem.

I think it is a real problem. The next question is where to patch this. 

> is replicated on all browsers(Chrome, Firefox and Edge), not only in Firefox.

The issue I linked is so old that situation may have changed since then

> because my problem is on vaadin-push 7.7.16

The behavior is most likely the same with Vaadin 8, since the patch proposed in the ticket I linked has not been merged to Atmosphere. Also I am not sure is that patch even correct one

See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI#Description

So the root cause of the problem is that encodeURI function does not even claim to work with IPv6:

_Also note that if one wishes to follow the more recent RFC3986 for URLs, which makes square brackets reserved (for IPv6) and thus not encoded when forming something which could be part of a URL (such as a host), the following code snippet may help:_

```
function fixedEncodeURI(str) {
    return encodeURI(str).replace(/%5B/g, '[').replace(/%5D/g, ']');
}
```

So the proper way to fix this is to apply the above to client part of the Atmosphere.

**Comment by petregeorgee** [comment]
I know that vaadin-push jar has a dependency on atmosphere-runtime jar.
From what I see, this jar has only java classes, not even one JavaScript
file.

How can be possible to include the fix on the atmosphere client? Corect me
if I am wrong.

I think the proper way to solve this issue is to include '[' and ']'
characters in the escaped characters list.

Thank you,
George

**Comment by petregeorgee** [comment]
I see that vaadin-pus has also a dependency of atmosphere-javascript. Now is very clear for me.  

I don't know why this: 
https://github.com/Atmosphere/atmosphere-javascript/pull/232
has never been merged.

**Comment by petregeorgee** [comment]
Hi @TatuLund, the fix was merged into the atmosphere's master branch. On which version of vaadin-push I could use the fix?

Thanks,
George

**Comment by petregeorgee** [comment]
It would be tremendous if the fix could be included on vaadin 7.7.x and newer versions.
